src/Common/src/System/Diagnostics/Debug.cs

```diff
@@ -84,19 +144,33 @@ public static void WriteLine(string message)
         [System.Diagnostics.Conditional("DEBUG")]
         public static void Write(string message)
         {
-            s_logger.WriteCore(message ?? string.Empty);
+            if (message == null)
+            {
+                WriteCore(string.Empty);
+                return;
+            }
+            if (s_needIndent)
+            {
+                message = GetIndentString() + message;
+                s_needIndent = false;
+            }
+            WriteCore(message);
+            if (message.EndsWith(Environment.NewLine))
+            {
+                s_needIndent = true;
```

Comment:
When I talked with @weshaggard we decided that it is fine if debug statements aren't always properly indented in multi-threaded scenarios. The code shouldn't crash, but indentation that is a bit off is fine. I don't know of any lock free way to make this completely correct.

The implementation in DefaultTraceListener also doesn't synchronize at all.